### ISSUE TITLE
Update the alphabetical sort requirement? 

### ISSUE THREAD

**Issue by cjerrington** [garritfra/darktheme.club#51]
Thanks for the PR (#50) to add my website to this list! With the change made on line 32 of the `index.md` file, we can have Jekyll sort the list on build for us as well. 

https://github.com/garritfra/darktheme.club/blob/76883d9bc9f05648e2fdf98cfb0a127aa686cb5a/index.md?plain=1#L32

With my first commit it failed since the `_data/sites.yml` file was not alphabetical as well when the GitHub Action ran to check the files. With the new change this could allow other individuals to add their content to the end of the files and add their on PR. When Jekyll builds, it'll sort the site for us too. 

If the `_data/sites.yml` file is still wanting to be sorted alphabetical, I found a few implementations of sorting `yaml` files on Jekyll builds or with a plugin. There could still be a shell script that could run as well.

**Comment by garritfra** [comment]
Thanks for your suggestion. I originally wanted to avoid unsorted lists, but I think you're right. There could be an action that opens a new PR for sorting the list whenever `main` is out of order.

Do you have time to investigate this?

**Comment by cjerrington** [comment]
I could take a look at a few options! I'll follow up when I have some research done for what might be best and automated.

**Comment by cjerrington** [comment]
If we want to keep the `_data/sites.yml` file alphabetized as well, I found a good processor: `yq`. This works similarly to how `jq` does and seems to better support yaml files. I've worked on the shell script to do this in the current action, then have the action add a commit if there is a change. 

The `yq` also has a [github action](https://mikefarah.gitbook.io/yq/usage/github-action) we could look at as well to process this. I'll do some more testing and make a request soon.

**Comment by cjerrington** [comment]
Here is what I've found. `yq` does what we need to. To keep things orderly, we have to parse and sort the `_data/sites.yml` file then commit changes and have the site built. We need to output the sorting to a new file, then delete the original and rename `_data/sorted.yml` to `_data/sites.yml`. 

`yq` does run on the action directly as well, or following the [documentation](https://mikefarah.gitbook.io/yq/usage/github-action) 

I am not the best with Github Actions, but this does work. The action works as followed.

```shell
yq 'sort_by(.domain)' _data/sites.yml > _data/sorted.yml
rm _data/sites.yml
mv _data/sorted.yml _data/sites.yml
git add .
git diff
git config --global user.email "github-action-bot@example.com"
git config --global user.name "GitHub Action Bot"
git commit -m "alphabatized _data/sites.yml" -a || echo "No changes to commit"
git push
```

The problem with this approach and the Continuous Integration and the pages-build-deployment is a new commit is added, the github-pages tries to build the site, then the workflow updates the `sites.yml` and cancels the deploy, sends an email. Then with the new commit from the Github Action Bot for the alpha sorting, github-pages kicks off and successfully deploys the site. 

If a commit comes in and the `sites.yml` is already alphabetized then github-pages continues successfully.  

Instead of the `pages-build-deployment` default action, a custom workflow action could be used it looks like as well. 

I think the solution works, just order of operations and the building of the site. The other option is to forgo the alpha sorting in the `_data/sites.yml` and still have Jekyll render the site alphabetically based on the `sort:"domain"`

**Comment by garritfra** [comment]
Sorry for my late response, and thanks for digging into this! This does look promising. There seems to be [an action](https://github.com/marketplace/actions/add-commit) dedicated to pushing changed files, which could come in handy here.

> The problem with this approach and the Continuous Integration and the pages-build-deployment is a new commit is added, the github-pages tries to build the site, then the workflow updates the sites.yml and cancels the deploy, sends an email.

I think we can circumvent this by creating a new action for this task, instead of adding it to the existing deploy action. So whenever a commit comes in (no matter if the build succeedes or not), the action checks if the domains are sorted, and if not, sorts them and pushes a new commit.

Some more thoughs: I think for now we might want to listen for commits only on `main` and open a new PR for fixing the alphabetical order, so that a maintainer actively has to come in and check if the action does what it's supposed to. If this proves to work, we could start sending those sort-commits straight to the open PRs to avoid additional manual labour.

Do you have some more time to spare for this? Otherwise, I'd be glad to help out once I find the time. 🙂

**Comment by cjerrington** [comment]
I think this is a promising solution as well. But I don't have the same level of access to manage the workflows like you do. I think default `pages-build-deploy` GitHub uses needs to change and the dependency of alpha sort on the pull request and merge. 

The process should be:
- user updates the `_data/sites.yml` as is
- custom action sorts the file, then deploys the built site, then merge the sorted file without causing another build of the site.

**Comment by Vinfall** [comment]
> custom action sorts the file, then deploys the built site, then merge the sorted file without causing another build of the site.

How does the build get triggered? GitHub action or Netlify?
If it's built by GitHub Action, all you need is to add a custom filter.

Change this:

https://github.com/garritfra/darktheme.club/blob/ee233ab0326452fa502a2cec72946fd97c81ebc6/.github/workflows/ci.yml#L9-L12

to something like this:

```yaml
jobs:
  ci:
    name: Check if domains are sorted alphabetically 
    runs-on: ubuntu-latest
    if: "! contains(github.event.head_commit.message, 'Site sort')"
```

And in the new action recipe, commit the sorted file with a message starting with `Site sort`.

**Comment by garritfra** [comment]
Fixed by #119. Thanks for your help and suggestions!
